Improve new contributor experience

mathrick opened the pull request:
This fixes a couple frustrations I ran into in my recent PRs:

- Adds a single `make` target to run all style checks that will be required for a PR
- Adds the missing dev dependency on `codespell`
- Updates `CONTRIBUTING.rst` to mention how to install Dulwich in dev mode, the required checks for PRs, and fixes an apparent typo in the coding style section (which seems to be the opposite of what the code actually uses)
- Fixes code blocks in `CONTRIBUTING.rst`  for proper rendering on GitHub

You might also want to consider using the `make` targets inside the PR test workflow instead of inlining their corresponding commands, to guarantee they won't drift apart, but that's a larger decision and I leave it up to you.

jelmer commented on CONTRIBUTING.rst:
```diff
@@ -6,12 +6,29 @@ Where possible include updates to NEWS along with your improvements.
+
+   $ cd ~/path/to/checkouts/dulwich
+   # Make sure to run this in a virtual environment
+   $ pip install -e ".[dev]"
```
This seems a little prescriptive - I've personally never used -e.

mathrick replied:
Sure, I can make it less prescriptive and more of a "here's an easy way to start".

mathrick replied:
I adjusted the wording to make it more of general outline, and expanded it to mention Rust / Cargo being required as well.

jelmer commented on CONTRIBUTING.rst:
```diff
@@ -6,12 +6,29 @@ Where possible include updates to NEWS along with your improvements.
+
 Coding style
 ------------
-Where possible, please follow PEP8 with regard to coding style. Run ruff.
```
This should just recommend running ruff without going into details.

mathrick replied:
Not sure what you mean, which part are you referring to? If the "PEP8" part, that's just the original contents of the file, I didn't change it :). I can update it however.

mathrick replied:
I adjusted it, does it look better to you?

jelmer commented on Makefile:
```diff
@@ -9,6 +9,8 @@ DESTDIR=/
 
 all: build
 
+all-style:: fix reformat codespell
```
style intentionally only reports on style, it doesn't make any changes. So it seems that all-style would.

fix makes actual code changes, it's not just a style thing.

mathrick replied:
I went ahead and added separate `all-style` and `all-fix`. I also reordered the targets within `Makefile` so they're grouped by area; this doesn't change any of the rules, just makes them easier to scan to find the one you're interested in IMHO.

mathrick commented:
I changed it to address your feedback. 

I also added `setuptools` and `setuptools-rust` to the `dev` requirements. I actually forgot that it was another snag I ran into immediately after cloning Dulwich, as `make check` requires them by virtue of invoking `setup.py`. Unfortunately, I don't know of a better way of signalling that `build-system.requires` dependencies are also needed here; I'm afraid listing them again is the only option.

mathrick commented:
Dulwich code itself doesn't need them, but all `make check-*` targets do, as they call `setup.py`, which requires both to be present. When it's through the `build-system.requires`, apparently it's enough to only list `setuptools` and it takes care of installing Rust extensions itself, but when it's `python setup.py`, then the same automatic dependency mechanism doesn't kick in and `setuptools-rust` needs to be listed as well.

jelmer commented on Makefile:
```diff
@@ -42,36 +54,34 @@ check-noextensions:: clean
+	ruff check --fix .
+
+format:
+	ruff format --check .
```
This is a change in behaviour, please just have "make format" do formatting - not just check whether reformatting is necessary.

mathrick replied:
But that's what `reformat` already does? This is a new target that only does checks, but doesn't change anything, to provide parity with other check/fix targets.

jelmer replied:
Oh, I see. Having both format and reformat where one makes changes confused me. Don't really want to bikeshed on this and happy to merge as-is, but I don't think it's a great name.

jelmer commented on Makefile:
```diff
@@ -7,20 +7,32 @@ COVERAGE = python3-coverage
 
-sphinx::
-	$(MAKE) -C docs html
+all-fix:: fix reformat codespell
```
codespell doesn't make changes so it seems odd to be included here.

I'd also argue that reformat does something fundamentally different from "ruff check --fix" which does more than just change formatting - it can affect behaviour.

mathrick replied:
That's news to me, and the [Ruff docs](https://docs.astral.sh/ruff/formatter/) make no mention of that, are you sure? The formatter is supposed to reformat the code only, not change how it parses. You might be thinking of `ruff check --unsafe-fixes`, which does say it could potentially change behaviour, but none of our rules use that.

jelmer replied:
Ruff is not meant to make changes to behaviour but it's possible, especially in corner cases. Unlikely "ruff format" I'd be uncomfortable running it unconditionally, for example.

mathrick replied:
I really think this would be considered a serious bug if `ruff format` changed any behaviour. IMHO splitting it out into separate rules that apply fixes, as opposed to simply checking, makes it pretty clear that that's their purpose. I don't think forcing people to apply all the mechanical formatting fixes manually to please the tool is an improvement, when the tool itself can apply them. This just extends the rules so that there's a comprehensive set of checks and matching fixes.

mathrick replied:
I also added `codespell-fix` to match the rest. Since spelling fixes are not necessarily obvious, I made it ask for confirmation before applying each one.

jelmer commented on Makefile:
```diff
@@ -7,20 +7,32 @@ COVERAGE = python3-coverage
 install::
 	$(SETUP) install --root="$(DESTDIR)"
 
+check-all: check check-pypy check-noextensions
```
This breaks when run with "make -j". 

"check-all" is a bit confusing name - what is this target really trying to do? "check" for most projects means running the test suite. I'm not sure why this also tests typing.

mathrick replied:
I don't know :). This is an existing rule, I just moved it around. I agree that it doesn't seem to have a clear purpose.

jelmer replied:
Let's just drop it?

mathrick replied:
Can do.

mathrick commented on CONTRIBUTING.rst:
```diff
@@ -6,12 +6,50 @@ Where possible include updates to NEWS along with your improvements.
+exact code style, please run it to make sure your changes are conformant. See also "Style
+and typing checks" below for details on running style checkers.
 
-Furthermore, triple-quotes should always be """, single quotes are ' unless
```
Sure, I can do that.

mathrick commented:
> > Dulwich code itself doesn't need them, but all `make check-*` targets do, as they call `setup.py`, which requires both to be present. When it's through the `build-system.requires`, apparently it's enough to only list `setuptools` and it takes care of installing Rust extensions itself, but when it's `python setup.py`, then the same automatic dependency mechanism doesn't kick in and `setuptools-rust` needs to be listed as well.
> 
> This is where it gets a bit muddy though - "pip install -e ." builds the rust bindings so when doing that, you probably don't want to recommend running make check at all, but just invoking the test suite directly (python -m unittest dulwich.tests.test_suite).

Hmm, that's a good point, but I also don't want to make people have to remember that. I'll see if I can make it work for both cases.

mathrick commented:
@jelmer: I added support for detecting editable install. I don't know what workflow you use yourself, so I wasn't able to test it very extensively, let me know if you run into any issues with it. For me it adds a little bit of overhead for TAB-completion, but I found it to be within acceptable limits. My setup is probably the slowest possible case, as I'm using `pyenv` and its shims are definitely not the fastest.

jelmer commented on Makefile:
```diff
@@ -4,33 +4,45 @@ SETUP = $(PYTHON) setup.py
 TESTRUNNER ?= unittest
 RUNTEST = PYTHONHASHSEED=random PYTHONPATH=$(shell pwd)$(if $(PYTHONPATH),:$(PYTHONPATH),) $(PYTHON) -m $(TESTRUNNER) $(TEST_OPTIONS)
 COVERAGE = python3-coverage
+EDITABLE := $(shell pip show dulwich 2>/dev/null | grep -E "^Editable project location:")
```
This doesn't work, it will suggest you're in an editable installation if there's a system-wide install of dulwich. 

pip maybe also not be available as "pip" but only as "pip3"

I'd keep this simple and avoid suggest running the Makefile to invoke the test suite in CONTRIBUTING.rst, but just running "python -m unittest dulwich.tests.test_suite".

jelmer replied:
On Debian systems where there's packages that use Dulwich this is quite common. This is the case on my own system, and running "pip show dulwich" here picks up the system version of Dulwich. This is breaking my personal workflow with the Makefile, because now it stops building the extensions.

I'd be inclined to suggest that CONTRIBUTING.rst doesn't reference the Makefile at all and just suggests running plain commands. That matches what's in the CI workflow, and addresses your concerns about the Make targets. And then CI can avoid touching the Makefile.

jelmer commented:
Hi @mathrick - I've trimmed this down a bit to exclude the Makefile changes, since I think the other changes to CONTRIBUTING.rst are really useful and I don't want to block those. I hope that's okay. Thanks for the PR!

Still open to refactoring of the makefile, but hopefully this will at least help new contributors.

mathrick commented:
> Hi @mathrick - I've trimmed this down a bit to exclude the Makefile changes, since I think the other changes to CONTRIBUTING.rst are really useful and I don't want to block those. I hope that's okay. Thanks for the PR!
> 
> Still open to refactoring of the makefile - and I appreciate this doesn't address all of your concerns (especially around CI being able to invoke the Makefile) but hopefully this will at least help new contributors.

Sure thing! Apologies for the late reply; I got pretty busy at work, so this will at least move things forward. I think splitting the make changes into their own thing is a very reasonable way of doing it.
